### ISSUE TITLE
torch.nn.functional.conv2d: unexpected mkl-dnn::impl::global_scratchpad_t::global_scratchpad_t(size_t) assertion failed

### ISSUE THREAD

**Issue by davtir** [pytorch/pytorch#31195]
Hi,
I'm experiencing some unexpected behaviours while using `torch.nn.function.conv2d` function, still doesn't know if it is actually a bug or something that I am missing.
In particular I got the following error while executing my project script:

`python: /pytorch/third_party/ideep/mkl-dnn/src/common/scratchpad.cpp:63: mkldnn::impl::global_scratchpad_t::global_scratchpad_t(size_t): Assertion scratchpad_ != nullptr failed.`

I was able to reproduce the error using the following simple code:

```
import torch
def main():
	dim = 181;
	input_data = torch.randn(1,1,dim ,dim );
	filter_data = torch.randn(1,1,dim ,dim );	
	torch.nn.functional.conv2d(filter_data, input_data, padding=(int(dim /2 - 1), int(dim /2 - 1)));
if __name__ == "__main__":
	main()
```
Notice that, in order to get the error dim must be >= 181 on my environment.

## Environment
```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: CentOS Linux release 7.6.1810 (Core)
GCC version: (GCC) 8.2.1 20180905 (Red Hat 8.2.1-3)
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla P100-PCIE-12GB
GPU 1: Tesla P100-PCIE-12GB
GPU 2: Tesla P100-PCIE-12GB
GPU 3: Tesla P100-PCIE-12GB

Nvidia driver version: 440.33.01
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect
```

## Additional context

Notice also that I tried to execute the above script on another machine with the following specifications:

```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect
```
I got the same results but with a different value for variable dim, in particular on this environment variable dim must be >=175 in order to reproduce the error.

Thank you in advance.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh

**Comment by colesbury** [comment]
I'm not seeing this error on my installation, but something might be different (like more RAM). It looks like `malloc` is failing:

Here's the relevant code.

```c++
struct global_scratchpad_t : public scratchpad_t {
    global_scratchpad_t(size_t size) {
        if (size > size_) {
            if (scratchpad_ != nullptr) free(scratchpad_);
            size_ = size;
            scratchpad_ = (char *) malloc(size, page_size);
            assert(scratchpad_ != nullptr);
        }
        reference_count_++;
    }
...
```

My guess is that MKL-DNN is trying to use a really big scratchpad because of poor algorithm selection. Instead of failing gracefully, it just dies with an assertion error.

So the fix might be two parts (both on the MKL-DNN side):
1) Don't use a huge scratchpad
2) Handle allocation failures gracefully

**Comment by davtir** [comment]
As additional info, the server on which the error occurs is equipped with 64Gb of RAM and has 2 Intel Xeon E5-2640 v4.

**Comment by colesbury** [comment]
Hmmm... when I run the example the whole program only uses ~4 GB. I'm running on a similar CPU.

**Comment by davtir** [comment]
I don't know which is the actual memory overhead introduced by mkl-dnn, but the memory used in the example is quite low... Maybe some buffer overflow causing malloc ti fail?

**Comment by terminator-ger** [comment]
I have the same issue on similar hardware, i also have a dual socket CPU with 2x Intel Xeon E5-2690 v4, 64GB RAM. Watching the memory consumption during runtime  i get nowhere near to the max available memory.

**Comment by XiaobingSuper** [comment]
@terminator-ger @davtir, thanks for report this issue, I will double checkout it.

**Comment by XiaobingSuper** [comment]
@terminator-ger @davtir , we will pull a request to sovle this bug soon. Thanks!

**Comment by f0k** [comment]
@XiaobingSuper: Thank you for looking into this! Could you use further input? I also got hit by the assertion, trying to convolve a 512x512 image with a 41x41 kernel and 20x20 padding. If it's hard to select a better algorithm, at least a Python exception would be great!

**Comment by XiaobingSuper** [comment]
@f0k, we have a patch to solved it internaly, and preparing to upgrade MKLDNN to 1.2 soon.  You can first apply the following patch to MKLDNN(cd third_party/ideep/mkl-dnn/ and than apply the patch)to solve this problem:
```
diff --git a/src/common/memory_tracking.hpp b/src/common/memory_tracking.hpp
index 9926d2b31..7106011bb 100644
--- a/src/common/memory_tracking.hpp
+++ b/src/common/memory_tracking.hpp
@@ -272,6 +272,8 @@ struct registrar_t {
             size_t alignment = default_alignment)
     { registry_.book(make_key(prefix_, key), size, alignment); }

+    size_t size() const { return registry_.size();  }
+
 protected:
     registry_t &registry_;
     const key_t prefix_;
diff --git a/src/cpu/gemm_convolution_utils.cpp b/src/cpu/gemm_convolution_utils.cpp
index b30f29385..311cf2911 100644
--- a/src/cpu/gemm_convolution_utils.cpp
+++ b/src/cpu/gemm_convolution_utils.cpp
@@ -1115,6 +1115,16 @@ status_t init_conf(jit_gemm_conv_conf_t &jcp,
                 sizeof(float) * jcp.oc * jcp.ngroups);
         }
     }
+    // Heuristic threshold for requested scratchpad memory to avoid
+    // possible crash on memory allocation:
+    // 1Gb or size of the buffers already used for this convolution proportional
+    // to the number of threads and multiplied by a heuristic coefficient (15)
+    size_t scratchpad_limit_by_absolute_value = (size_t)1 << 30; // 1Gb
+    size_t scratchpad_limit_by_tensor_sizes = 15 * max_threads
+            * (src_d.size() + weights_d.size() + dst_d.size());
+    size_t scratchpad_limit = nstl::min(scratchpad_limit_by_absolute_value,
+            scratchpad_limit_by_tensor_sizes);
+    if (scratchpad.size() > scratchpad_limit) return status::unimplemented;
     return status::success;
 }
```
